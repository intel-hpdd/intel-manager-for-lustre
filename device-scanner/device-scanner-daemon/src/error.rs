// Copyright (c) 2019 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use futures::channel::mpsc::TrySendError;
use std::{error, fmt, io, num, result};
use tokio::codec::LinesCodecError;

pub type Result<T> = result::Result<T, Error>;

pub fn none_error<E>(error: E) -> Error
where
    E: Into<Box<dyn error::Error + Send + Sync>>,
{
    Error::NoneError(error.into())
}

#[derive(Debug)]
pub enum Error {
    Io(io::Error),
    TrySendError(Box<dyn error::Error + Send>),
    SerdeJson(serde_json::Error),
    LinesCodecError(LinesCodecError),
    LibZfsError(libzfs_types::LibZfsError),
    ParseIntError(num::ParseIntError),
    NoneError(Box<dyn error::Error + Send + Sync>),
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match *self {
            Error::Io(ref err) => write!(f, "{}", err),
            Error::TrySendError(ref err) => write!(f, "{}", err),
            Error::SerdeJson(ref err) => write!(f, "{}", err),
            Error::LinesCodecError(ref err) => write!(f, "{}", err),
            Error::LibZfsError(ref err) => write!(f, "{}", err),
            Error::ParseIntError(ref err) => write!(f, "{}", err),
            Error::NoneError(ref err) => write!(f, "{}", err),
        }
    }
}

impl error::Error for Error {
    fn cause(&self) -> Option<&dyn error::Error> {
        match *self {
            Error::Io(ref err) => Some(err),
            Error::TrySendError(_) => None,
            Error::SerdeJson(ref err) => Some(err),
            Error::LinesCodecError(ref err) => Some(err),
            Error::LibZfsError(ref err) => Some(err),
            Error::ParseIntError(ref err) => Some(err),
            Error::NoneError(_) => None,
        }
    }
}

impl From<io::Error> for Error {
    fn from(err: io::Error) -> Self {
        Error::Io(err)
    }
}

impl From<LinesCodecError> for Error {
    fn from(err: LinesCodecError) -> Self {
        Error::LinesCodecError(err)
    }
}

impl<E> From<TrySendError<E>> for Error
where
    E: Send + 'static,
{
    fn from(err: TrySendError<E>) -> Self {
        Error::TrySendError(Box::new(err))
    }
}

impl From<serde_json::Error> for Error {
    fn from(err: serde_json::Error) -> Self {
        Error::SerdeJson(err)
    }
}

impl From<libzfs_types::LibZfsError> for Error {
    fn from(err: libzfs_types::LibZfsError) -> Self {
        Error::LibZfsError(err)
    }
}

impl From<num::ParseIntError> for Error {
    fn from(err: num::ParseIntError) -> Self {
        Error::ParseIntError(err)
    }
}
