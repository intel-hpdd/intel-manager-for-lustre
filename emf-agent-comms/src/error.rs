// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use emf_rabbit::EmfRabbitError;
use futures::channel::oneshot;
use warp::reject;

#[derive(Debug)]
pub enum EmfAgentCommsError {
    EmfRabbitError(EmfRabbitError),
    SerdeJsonError(serde_json::error::Error),
    OneshotCanceled(oneshot::Canceled),
}

impl reject::Reject for EmfAgentCommsError {}

impl std::fmt::Display for EmfAgentCommsError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match *self {
            EmfAgentCommsError::EmfRabbitError(ref err) => write!(f, "{}", err),
            EmfAgentCommsError::SerdeJsonError(ref err) => write!(f, "{}", err),
            EmfAgentCommsError::OneshotCanceled(ref err) => write!(f, "{}", err),
        }
    }
}

impl std::error::Error for EmfAgentCommsError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match *self {
            EmfAgentCommsError::EmfRabbitError(ref err) => Some(err),
            EmfAgentCommsError::SerdeJsonError(ref err) => Some(err),
            EmfAgentCommsError::OneshotCanceled(ref err) => Some(err),
        }
    }
}

impl From<EmfRabbitError> for EmfAgentCommsError {
    fn from(err: EmfRabbitError) -> Self {
        EmfAgentCommsError::EmfRabbitError(err)
    }
}

impl From<serde_json::error::Error> for EmfAgentCommsError {
    fn from(err: serde_json::error::Error) -> Self {
        EmfAgentCommsError::SerdeJsonError(err)
    }
}

impl From<oneshot::Canceled> for EmfAgentCommsError {
    fn from(err: oneshot::Canceled) -> Self {
        EmfAgentCommsError::OneshotCanceled(err)
    }
}
