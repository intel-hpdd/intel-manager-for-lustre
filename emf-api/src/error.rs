// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use emf_manager_client::EmfManagerClientError;
use futures::channel::oneshot;
use thiserror::Error;
use warp::reject;

#[derive(Debug, Error)]
pub enum EmfApiError {
    #[error(transparent)]
    EmfManagerClientError(#[from] EmfManagerClientError),
    #[error("Not Found")]
    NoneError,
    #[error(transparent)]
    OneshotCanceled(#[from] oneshot::Canceled),
    #[error(transparent)]
    SerdeJsonError(#[from] serde_json::error::Error),
    #[error(transparent)]
    SqlxError(#[from] sqlx::Error),
    #[error("Filesystem Not Found")]
    FilesystemNotFound,
    #[error("Filesystem Not Found")]
    MgsNotFound,
}

impl reject::Reject for EmfApiError {}
