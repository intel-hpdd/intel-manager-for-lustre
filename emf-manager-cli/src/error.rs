// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use emf_wire_types::Command;
use thiserror::Error;

#[derive(Debug)]
pub enum DurationParseError {
    NoUnit,
    InvalidUnit,
    InvalidValue,
}

#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
#[serde(rename_all = "snake_case")]
pub enum RunStratagemCommandResult {
    DurationOrderError,
    FilesystemRequired,
    FilesystemDoesNotExist,
    FilesystemUnavailable,
    InvalidArgument,
    PurgeDurationTooBig,
    ReportDurationTooBig,
    PurgeDurationTooSmall,
    ReportDurationTooSmall,
    Mdt0NotFound,
    Mdt0NotMounted,
    StratagemServerProfileNotInstalled,
    StratagemClientProfileNotInstalled,
    RequiredFieldsMissing,
    ServerError,
    UnknownError,
}

#[derive(serde::Serialize, serde::Deserialize, Debug)]
pub struct RunStratagemValidationError {
    pub code: RunStratagemCommandResult,
    pub message: String,
}

impl std::fmt::Display for DurationParseError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            DurationParseError::NoUnit => write!(f, "No unit specified."),
            DurationParseError::InvalidUnit => {
                write!(f, "Invalid unit. Valid units include 'h' and 'd'.")
            }
            DurationParseError::InvalidValue => {
                write!(f, "Invalid value specified. Must be a valid integer.")
            }
        }
    }
}

impl std::error::Error for DurationParseError {}

#[derive(Debug, Error)]
pub enum EmfManagerCliError {
    ApiError(String),
    ClientRequestError(#[from] emf_manager_client::EmfManagerClientError),
    CmdUtilError(#[from] emf_command_utils::CmdUtilError),
    CombineEasyError(combine::stream::easy::Errors<char, &'static str, usize>),
    DoesNotExist(&'static str),
    FailedCommandError(Vec<Command>),
    FromUtf8Error(#[from] std::string::FromUtf8Error),
    Infallible(#[from] std::convert::Infallible),
    EmfGraphqlQueriesErrors(#[from] emf_graphql_queries::Errors),
    IntParseError(#[from] std::num::ParseIntError),
    IoError(#[from] std::io::Error),
    ParseDurationError(#[from] DurationParseError),
    ReqwestError(#[from] reqwest::Error),
    RunStratagemValidationError(#[from] RunStratagemValidationError),
    SerdeJsonError(#[from] serde_json::error::Error),
    TokioJoinError(#[from] tokio::task::JoinError),
    TokioTimerError(#[from] tokio::time::Error),
}

impl std::fmt::Display for EmfManagerCliError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match *self {
            EmfManagerCliError::ApiError(ref err) => write!(f, "{}", err),
            EmfManagerCliError::ClientRequestError(ref err) => write!(f, "{}", err),
            EmfManagerCliError::CmdUtilError(ref err) => write!(f, "{}", err),
            EmfManagerCliError::CombineEasyError(ref err) => write!(f, "{}", err),
            EmfManagerCliError::DoesNotExist(ref err) => write!(f, "{} does not exist", err),
            EmfManagerCliError::FailedCommandError(ref xs) => {
                let failed_msg = xs.iter().fold(
                    String::from("The following commands have failed:\n"),
                    |acc, x| format!("{}{}\n", acc, x.message),
                );

                write!(f, "{}", failed_msg)
            }
            EmfManagerCliError::FromUtf8Error(ref err) => write!(f, "{}", err),
            EmfManagerCliError::EmfGraphqlQueriesErrors(ref err) => write!(f, "{}", err),
            EmfManagerCliError::Infallible(_) => {
                write!(f, "A (supposedly) impossible situation has occurred")
            }
            EmfManagerCliError::IntParseError(ref err) => write!(f, "{}", err),
            EmfManagerCliError::IoError(ref err) => write!(f, "{}", err),
            EmfManagerCliError::ParseDurationError(ref err) => write!(f, "{}", err),
            EmfManagerCliError::ReqwestError(ref err) => write!(f, "{}", err),
            EmfManagerCliError::RunStratagemValidationError(ref err) => write!(f, "{}", err),
            EmfManagerCliError::SerdeJsonError(ref err) => write!(f, "{}", err),
            EmfManagerCliError::TokioJoinError(ref err) => write!(f, "{}", err),
            EmfManagerCliError::TokioTimerError(ref err) => write!(f, "{}", err),
        }
    }
}

impl std::fmt::Display for RunStratagemValidationError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl std::error::Error for RunStratagemValidationError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }
}

impl From<combine::stream::easy::Errors<char, &str, usize>> for EmfManagerCliError {
    fn from(err: combine::stream::easy::Errors<char, &str, usize>) -> Self {
        EmfManagerCliError::CombineEasyError(err.map_range(|_| ""))
    }
}

impl From<Vec<Command>> for EmfManagerCliError {
    fn from(xs: Vec<Command>) -> Self {
        EmfManagerCliError::FailedCommandError(xs)
    }
}
