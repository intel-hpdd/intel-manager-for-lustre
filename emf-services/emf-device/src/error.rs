// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use thiserror::Error;

#[derive(Error, Debug)]
pub enum EmfDeviceError {
    #[error(transparent)]
    SerdeJsonError(#[from] serde_json::Error),
    #[error(transparent)]
    SqlxCoreError(#[from] sqlx::Error),
    #[error(transparent)]
    SqlxMigrateError(#[from] sqlx::migrate::MigrateError),
    #[error("Could not find index in target name {0}")]
    TargetIndexNotFoundError(String),
}
