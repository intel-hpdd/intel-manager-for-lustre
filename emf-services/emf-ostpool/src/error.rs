// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("Not Found")]
    NotFound,
    #[error(transparent)]
    SqlxError(#[from] emf_postgres::sqlx::Error),
}
