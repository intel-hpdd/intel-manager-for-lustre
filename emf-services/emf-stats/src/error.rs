// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use thiserror::Error;

#[derive(Debug, Error)]
pub enum EmfStatsError {
    #[error(transparent)]
    EmfInfluxError(#[from] emf_influx::Error),
    #[error(transparent)]
    SystemTimeError(#[from] std::time::SystemTimeError),
}
