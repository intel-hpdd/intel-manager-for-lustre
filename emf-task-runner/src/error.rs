// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use thiserror::Error;

#[derive(Debug, Error)]
pub enum EmfTaskRunnerError {
    #[error(transparent)]
    JsonError(#[from] serde_json::Error),
    #[error(transparent)]
    SqlxError(#[from] sqlx::Error),
}
