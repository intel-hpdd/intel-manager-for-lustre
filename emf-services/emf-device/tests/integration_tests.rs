// Copyright (c) 2021 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

extern crate emf_device;

use chrono::{DateTime, Utc};
use device_types::{
    mount::{FsType, Mount, MountOpts, MountPoint},
    DevicePath,
};
use emf_device::update_client_mounts;
use emf_postgres::sqlx;
use emf_wire_types::Fqdn;
use insta::assert_json_snapshot;
use std::error::Error;

#[derive(serde::Serialize)]
struct ClientMount {
    id: i32,
    state_modified_at: DateTime<Utc>,
    state: String,
    filesystem: String,
    host_id: i32,
    mountpoints: Vec<String>,
}

#[tokio::test]
#[ignore = "Requires an active and populated DB"]
async fn test_insert() -> Result<(), Box<dyn Error>> {
    let pool = emf_postgres::test_setup().await?;

    sqlx::query!(
        r#"INSERT INTO host
        (
            machine_id,
            state,
            fqdn,
            boot_time
        ) VALUES
        ('1234', 'unconfigured', 'foo.bar', now())
        ON CONFLICT DO NOTHING"#
    )
    .execute(&pool)
    .await?;

    let mut mounts = im::HashSet::new();

    mounts.insert(Mount::new(
        MountPoint("/mnt/part1".into()),
        DevicePath("/dev/sde1".into()),
        FsType("ext4".to_string()),
        MountOpts("rw,relatime,data=ordered".to_string()),
    ));

    mounts.insert(Mount::new(
        MountPoint("/mnt/fs0a9c".into()),
        DevicePath("172.60.0.40@o2ib,172.60.0.44@o2ib:172.60.0.42@o2ib,172.60.0.46@o2ib:172.60.0.43@o2ib,172.60.0.47@o2ib:172.60.0.41@o2ib,172.60.0.45@o2ib:/fs0a9c".into()),
        FsType("lustre".into()),
        MountOpts("rw,flock,lazystatfs".into())
    ));

    mounts.insert(Mount::new(
        MountPoint("/mnt/fs0a9c2".into()),
        DevicePath("172.60.0.40@o2ib,172.60.0.44@o2ib:172.60.0.42@o2ib,172.60.0.46@o2ib:172.60.0.43@o2ib,172.60.0.47@o2ib:172.60.0.41@o2ib,172.60.0.45@o2ib:/fs0a9c".into()),
        FsType("lustre".into()),
        MountOpts("rw,flock,lazystatfs".into())
    ));

    // Add some mounts
    update_client_mounts(&pool, &Fqdn("foo.bar".into()), &mounts).await?;

    let xs = sqlx::query_as!(ClientMount, r#"SELECT * FROM lustreclientmount"#)
        .fetch_all(&pool)
        .await?;

    assert_json_snapshot!(xs, {
        "[].id" => "<ID>",
        "[].state_modified_at" => "<STATE_MODIFIED_AT>",
        "[].host_id" => "<HOST_ID>"
    });

    // Remove Some mounts
    update_client_mounts(&pool, &Fqdn("foo.bar".into()), &im::hashset![Mount::new(
        MountPoint("/mnt/fs0a9c2".into()),
        DevicePath("172.60.0.40@o2ib,172.60.0.44@o2ib:172.60.0.42@o2ib,172.60.0.46@o2ib:172.60.0.43@o2ib,172.60.0.47@o2ib:172.60.0.41@o2ib,172.60.0.45@o2ib:/fs0a9c".into()),
        FsType("lustre".into()),
        MountOpts("rw,flock,lazystatfs".into())
    )]).await?;

    let xs = sqlx::query_as!(ClientMount, r#"SELECT * FROM lustreclientmount"#)
        .fetch_all(&pool)
        .await?;

    assert_json_snapshot!(xs, {
        "[].id" => "<ID>",
        "[].state_modified_at" => "<STATE_MODIFIED_AT>",
        "[].host_id" => "<HOST_ID>"
    });

    // Remove all mounts
    update_client_mounts(&pool, &Fqdn("foo.bar".into()), &im::hashset![]).await?;

    let xs = sqlx::query_as!(ClientMount, r#"SELECT * FROM lustreclientmount"#)
        .fetch_all(&pool)
        .await?;

    assert_json_snapshot!(xs, {
        "[].id" => "<ID>",
        "[].state_modified_at" => "<STATE_MODIFIED_AT>",
        "[].host_id" => "<HOST_ID>"
    });

    Ok(())
}
