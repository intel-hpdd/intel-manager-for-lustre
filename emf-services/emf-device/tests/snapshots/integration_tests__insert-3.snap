---
source: emf-services/emf-device/tests/integration_tests.rs
expression: xs

---
[
  {
    "id": "<ID>",
    "state_modified_at": "<STATE_MODIFIED_AT>",
    "state": "unmounted",
    "filesystem": "fs0a9c",
    "host_id": "<HOST_ID>",
    "mountpoints": []
  }
]
